dev/PushNotifications/PushNotificationReceivedEventArgs.cpp

```diff
@@ -0,0 +1,118 @@
+#include "Microsoft.Windows.PushNotifications.PushNotificationReceivedEventArgs.g.cpp"
+#include <iostream>
+
+namespace winrt
+{
+    using namespace Windows::ApplicationModel::Background;
+    using namespace Windows::Storage::Streams;
+    using namespace Windows::Networking::PushNotifications;
+}
+
+namespace winrt::Microsoft::Windows::PushNotifications::implementation
+{
+    PushNotificationReceivedEventArgs::PushNotificationReceivedEventArgs(winrt::IBackgroundTaskInstance const& backgroundTask)
+    {
+        m_backgroundTaskInstance = backgroundTask;
+
+        // Need to process the RawNotification payload here because TriggerDetails is set to nullptr leaving IBackgroundTaskInstance::Run 
+        auto triggerDetails = m_backgroundTaskInstance.TriggerDetails();
+        RawNotification rawNotification = triggerDetails.as<RawNotification>();
+
+        IBuffer rawContentAsBuffer = rawNotification.ContentBytes();
+        DataReader dataReader = DataReader::FromBuffer(rawContentAsBuffer);
+
+        m_backgroundPayload = winrt::com_array<uint8_t>(rawContentAsBuffer.Length());
+        dataReader.ReadBytes(m_backgroundPayload);
+    }
+
+    PushNotificationReceivedEventArgs::PushNotificationReceivedEventArgs(winrt::PushNotificationReceivedEventArgs const& args)
+    {
+        m_args = args;
+    }
+
+    winrt::Microsoft::Windows::PushNotifications::PushNotificationReceivedEventArgs PushNotificationReceivedEventArgs::CreateFromBackgroundTaskInstance(winrt::IBackgroundTaskInstance const& backgroundTask)
+    {
+        return winrt::make<PushNotificationReceivedEventArgs>(backgroundTask);
+    }
+
+    winrt::Microsoft::Windows::PushNotifications::PushNotificationReceivedEventArgs PushNotificationReceivedEventArgs::CreateFromPushNotificationReceivedEventArgs(winrt::PushNotificationReceivedEventArgs const& args)
+    {
+        return winrt::make<PushNotificationReceivedEventArgs>(args);
+    }
+
+    winrt::com_array<uint8_t> PushNotificationReceivedEventArgs::Payload()
+    {
+        auto lock = m_lock.lock_shared();
+        winrt::com_array<uint8_t> payload{};
+        if (m_args)
+        {
+            RawNotification rawNotification = m_args.RawNotification();
+
+            IBuffer rawContentAsBuffer = rawNotification.ContentBytes();
+            DataReader dataReader = DataReader::FromBuffer(rawContentAsBuffer);
+
+            payload = winrt::com_array<uint8_t>(rawContentAsBuffer.Length());
+            dataReader.ReadBytes(payload);
+        }
+        else if (m_backgroundTaskInstance)
+        {
+            return winrt::com_array<uint8_t>(m_backgroundPayload.begin(), m_backgroundPayload.end());
+        }
+        return payload;
+    }
+
+    winrt::BackgroundTaskDeferral PushNotificationReceivedEventArgs::GetDeferral()
+    {
+        auto lock = m_lock.lock_shared();
+        if (m_backgroundTaskInstance)
+        {
+            return m_backgroundTaskInstance.GetDeferral();
+        }
+        return nullptr;
```

Comment:
Should it be an error to ask for a deferral from something that doesn't have one?

Returning `nullptr` here means that the caller is just going to crash later when they try to complete the null deferral.

Reply:
What HRESULT should be thrown here?

Reply:
Maybe `throw hresult_illegal_method_call()`?